File: backend/routers/game.py
from fastapi import APIRouter, HTTPException

from game.state_manager import get_session, update_state
from game.game_engine import (
    is_adjacent,
    apply_action_result,
    apply_research_result,
    advance_turn,
    check_victory,
)
from game.models import ActionRequest, ResearchRequest
from game.map_data import COUNTRIES
from ai.llm_client import resolve_attack, resolve_research
from routers.ws import broadcast

router = APIRouter(prefix="/game", tags=["game"])


@router.get("/map")
async def get_map():
    """Return the static Europe map data (countries, adjacency, default stats)."""
    return COUNTRIES


@router.get("/{game_id}/state")
async def get_state(game_id: str):
    """Return the full current game state."""
    state = get_session(game_id)
    if not state:
        raise HTTPException(status_code=404, detail="Game not found.")
    return state


@router.post("/{game_id}/action")
async def perform_action(game_id: str, body: ActionRequest):
    """
    Perform an attack or diplomacy action against an adjacent country.
    This is the player's only action for this turn; the turn advances afterwards.
    """
    state = get_session(game_id)
    if not state:
        raise HTTPException(status_code=404, detail="Game not found.")
    if state.status != "active":
        raise HTTPException(status_code=400, detail="Game is not active.")
    if state.current_turn_player_id != body.player_id:
        raise HTTPException(status_code=403, detail="It is not your turn.")

    player = next((p for p in state.players if p.player_id == body.player_id), None)
    if not player:
        raise HTTPException(status_code=404, detail="Player not found.")
    if player.has_acted_this_turn:
        raise HTTPException(status_code=400, detail="You have already acted this turn.")

    if not is_adjacent(state, body.player_id, body.attack):
        raise HTTPException(
            status_code=400,
            detail=f"'{body.attack}' is not adjacent to your empire.",
        )

    # Validate items_used are in player's inventory
    for item in body.items_used:
        if not any(r.item_name == item.item_name and r.quantity > 0 for r in player.resources):
            raise HTTPException(
                status_code=400, detail=f"Item not in inventory: {item.item_name}"
            )

    target = state.countries[body.attack]
    story_context = " ".join(state.story_log[-3:]) if state.story_log else ""

    action_response = await resolve_attack(
        attacker_name=player.name,
        attacker_army=player.army_strength,
        target_country=body.attack,
        target_army=target.army_strength,
        target_morale=target.morale,
        attack_type=body.attack_type,
        items_used=[item.model_dump() for item in body.items_used],
        story_context=story_context,
    )

    player.has_acted_this_turn = True
    state = apply_action_result(state, action_response, body.player_id, body.attack, body.items_used)

    winner = check_victory(state)
    if winner:
        state.status = "finished"
        state.winner_id = winner
        update_state(game_id, state)
        await broadcast(
            game_id,
            {"type": "game_over", "winner_id": winner, "state": state.model_dump(mode="json")},
        )
    else:
        state = advance_turn(state)
        update_state(game_id, state)
        await broadcast(
            game_id,
            {"type": "state_update", "state": state.model_dump(mode="json")},
        )

    return action_response


@router.post("/{game_id}/research")
async def perform_research(game_id: str, body: ResearchRequest):
    """
    Research an item or technology.
    This is the player's only action for this turn; the turn advances afterwards.
    """
    state = get_session(game_id)
    if not state:
        raise HTTPException(status_code=404, detail="Game not found.")
    if state.status != "active":
        raise HTTPException(status_code=400, detail="Game is not active.")
    if state.current_turn_player_id != body.player_id:
        raise HTTPException(status_code=403, detail="It is not your turn.")

    player = next((p for p in state.players if p.player_id == body.player_id), None)
    if not player:
        raise HTTPException(status_code=404, detail="Player not found.")
    if player.has_acted_this_turn:
        raise HTTPException(status_code=400, detail="You have already acted this turn.")

    # Validate resources_used are in player's inventory
    for resource_name in body.resources_used:
        if not any(r.item_name == resource_name and r.quantity > 0 for r in player.resources):
            raise HTTPException(
                status_code=400, detail=f"Resource not in inventory: {resource_name}"
            )

    research_response = await resolve_research(
        player_name=player.name,
        item=body.item,
        resources_used=body.resources_used,
    )

    player.has_acted_this_turn = True
    state = apply_research_result(state, research_response, body.player_id, body.resources_used)
    state = advance_turn(state)
    update_state(game_id, state)

    await broadcast(
        game_id,
        {"type": "state_update", "state": state.model_dump(mode="json")},
    )

    return research_response

File: backend/src/routes/game.ts
import { Router, Request, Response } from "express";
import { getSession, updateState } from "../game/stateManager";
import {
  isAdjacent,
  applyActionResult,
  applyDiplomacyResult,
  applyResearchResult,
  advanceTurn,
  checkVictory,
} from "../game/gameEngine";
import {
  ActionRequest,
  Country,
  GameState,
  Player,
  ResearchRequest,
} from "../game/models";
import { startCountries } from "../game/mapData";
import {
  resolveAttack,
  resolveDiplomacy,
  resolveResearch,
} from "../ai/llmClient";
import { broadcastState, broadcastGameOver } from "../ws/wsServer";

export const gameRoutes = Router();

const validateState = (
  state: GameState | undefined,
  body: { playerId: string },
  res: Response,
): state is GameState => {
  if (!state) {
    res.status(404).json({ error: "Game not found." });
    return false;
  }

  if (state.status !== "active") {
    res.status(400).json({ error: "Game is not active." });
    return false;
  }

  if (state.currentTurnPlayerId !== body.playerId) {
    res.status(403).json({ error: "It is not your turn." });
    return false;
  }

  return true;
};

export const getMap = async (req: Request, res: Response) => {
  res.json(startCountries);
};

export const getState = async (req: Request, res: Response) => {
  const gameId = req.params.gameId;
  const state = getSession(gameId);

  if (!state) {
    res.status(404).json({ error: "Game not found." });
    return;
  }
  res.json(state);
};

// Handle player action (attack or diplomacy)
export const performAction = async (req: Request, res: Response) => {
  const gameId = req.params.gameId;
  const body: ActionRequest = req.body;
  const state = getSession(gameId);
  console.log(body);

  if (!validateState(state, body, res)) {
    return;
  }

  const player: Player | undefined = state.players.find(
    (player) => player.playerId === body.playerId,
  );
  console.log(state);

  console.log(player);

  // Validate the action request
  if (!player) {
    res.status(404).json({ error: "Player not found." });
    return;
  }

  if (player.hasActedThisTurn) {
    res.status(400).json({ error: "You have already acted this turn." });
    return;
  }

  if (!isAdjacent(state, body.playerId, body.target)) {
    res.status(400).json({
      error: `'${body.target}' is not adjacent to your empire.`,
    });
    return;
  }

  // Validate itemsUsed are in player's inventory
  const missingItems = body.itemsUsed
    .map((item) =>
      !player.resources.some(
        (r) => r.resourceName === item.itemName && r.quantity > 0,
      )
        ? item.itemName
        : null,
    )
    .filter(Boolean);

  if (missingItems.length > 0) {
    res.status(400).json({
      error: `Item not in inventory: ${missingItems[0]}`,
    });
    return;
  }

  // Get the target country story history, items, and prepare for action resolution
  const target: Country = state.countries[body.target];
  const storyContext = state.storyLog.slice(-3).join(" ");
  const mappedItems = body.itemsUsed.map((item) => ({
    resourceName: item.itemName,
    resourceEffect: item.itemEffect,
    quantity: 1,
  }));

  player.hasActedThisTurn = true;

  // Resolve the action based on the attackType (war or diplomatic)
  const handlers = {
    // TODO: resoleDiplomacy takes a Country as parameter. Now Player has no Country
    // Should whole Player schema be redesign, or should all data come from body?
    // Also should Player have country that is one big country where all new countries gets merged into
    diplomatic: async () => {
      const outcome = await resolveDiplomacy(
        player.empire,
        target,
        mappedItems,
      );

      return {
        response: outcome,
        newState: applyDiplomacyResult(
          state,
          outcome,
          body.playerId,
          body.target,
          body.itemsUsed,
        ),
      };
    },
    // Resolve Attack
    war: async () => {
      const outcome = await resolveAttack(
        player.empire,
        player.armyStrength,
        player.morale,
        target,
        mappedItems,
        storyContext,
      );

      return {
        response: outcome,
        newState: applyActionResult(
          state,
          outcome,
          body.playerId,
          body.target,
          body.itemsUsed,
        ),
      };
    },
  };

  const { response, newState } = await handlers[body.attackType]();

  const winner = checkVictory(newState);
  if (winner) {
    newState.status = "finished";
    newState.winnerId = winner;
  } else {
    advanceTurn(newState);
  }

  updateState(gameId, newState);
  if (winner) {
    broadcastGameOver(gameId);
  } else {
    broadcastState(gameId);
  }
  res.json(response);
};

export const performResearch = async (req: Request, res: Response) => {
  const gameId = req.params.gameId;
  const body: ResearchRequest = req.body;

  const state = getSession(gameId);
  if (!validateState(state, body, res)) {
    return;
  }

  const player = state.players.find((p) => p.playerId === body.playerId);
  if (!player) {
    res.status(404).json({ error: "Player not found." });
    return;
  }
  if (player.hasActedThisTurn) {
    res.status(400).json({ error: "You have already acted this turn." });
    return;
  }

  // Validate itemsUsed are in player's inventory
  const missingItems = body.resourcesUsed
    .map((resource) =>
      !player.resources.some(
        (r) => r.resourceName === resource && r.quantity > 0,
      )
        ? resource
        : null,
    )
    .filter(Boolean);

  if (missingItems.length > 0) {
    res.status(400).json({
      error: `Item not in inventory: ${missingItems[0]}`,
    });
    return;
  }

  const researchResponse = await resolveResearch(
    player.name,
    body.item,

    body.resourcesUsed.map((name) => {
      const resource = player.resources.find(
        (resource) => resource.resourceName === name,
      );

      return {
        resourceName: resource?.resourceName ?? name,
        resourceEffect: resource?.resourceEffect ?? "",
        quantity: 1,
      };
    }),
  );

  player.hasActedThisTurn = true;
  const newState = applyResearchResult(
    state,
    researchResponse,
    body.playerId,
    body.resourcesUsed,
  );

  advanceTurn(newState);
  updateState(gameId, newState);
  broadcastState(gameId);
  res.json(researchResponse);
};

gameRoutes.get("/game/map", getMap);
gameRoutes.get("/game/:gameId/state", getState);
gameRoutes.post("/game/:gameId/action", performAction);
gameRoutes.post("/game/:gameId/research", performResearch);
